File: apps/server-java/src/main/java/com/maisonamane/port/driven/repository/PilotProductRepository.java
package com.maisonamane.port.driven.repository;

import com.maisonamane.domain.pilot.aggregate.PilotProduct;
import com.maisonamane.domain.pilot.valueobject.ProductId;
import com.maisonamane.port.driven.error.PersistenceError;
import io.vavr.control.Either;
import io.vavr.control.Option;

/**
 * Repository interface for PilotProduct (Write Model).
 * Maps Effect-TS Context.Tag interface to Java interface with Either returns.
 *
 * Effect<T, E> → Either<E, T>
 * Option<T> → io.vavr.control.Option<T>
 */
public interface PilotProductRepository {

    Either<PersistenceError, PilotProduct> save(PilotProduct product);

    Either<PersistenceError, Option<PilotProduct>> findById(ProductId id);

    Either<PersistenceError, PilotProduct> update(PilotProduct product);
}

File: apps/server/src/ports/driven/repositories/pilot-product.repository.ts
// src/ports/driven/repositories/pilot-product.repository.ts

import { Context } from 'effect'
import type { Effect } from 'effect/Effect'
import type { Option } from 'effect/Option'

import type { ProductId, PilotProduct } from '../../../domain/pilot'

import type { PersistenceError } from './errors'

// ============================================
// PILOT PRODUCT REPOSITORY (Write Model)
// ============================================

export interface PilotProductRepositoryService {
  readonly save: (product: PilotProduct) => Effect<PilotProduct, PersistenceError>

  readonly update: (product: PilotProduct) => Effect<PilotProduct, PersistenceError>

  readonly findById: (id: ProductId) => Effect<Option<PilotProduct>, PersistenceError>
}

export class PilotProductRepository extends Context.Tag('PilotProductRepository')<
  PilotProductRepository,
  PilotProductRepositoryService
>() {}
